k8srm-prototype/pkg/gen/nvidia.go

```diff
@@ -0,0 +1,159 @@
+	allocatable, err := l.GetPerGpuAllocatableDevices(0)
+	if err != nil {
+		return nil, err
+	}
+
+	model := newresourceapi.PerGpuAllocatableDevices(allocatable).ToNamedResourcesResourceModel()
+
+	if model.NamedResources == nil {
+		return nil, fmt.Errorf("found no named resource object in the model")
+	}
+
+	if len(model.NamedResources.SharedLimits) != 1 {
+		return nil, fmt.Errorf("found %d shared limit groups in the resources", len(model.NamedResources.SharedLimits))
+	}
+
+	var devices []api.Device
+	for _, instance := range model.NamedResources.Instances {
+		devices = append(devices, instanceToDevice(instance))
+	}
+
+	return &api.DevicePool{
+		TypeMeta: metav1.TypeMeta{
+			APIVersion: api.DevMgmtAPIVersion,
+			Kind:       "DevicePool",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name: nodeName + "-" + poolName,
+		},
+		Spec: api.DevicePoolSpec{
+			NodeName: &nodeName,
+			Driver:   "gpu.nvidia.com/dra",
+			Attributes: []api.Attribute{
+				{Name: "vendor", StringValue: ptr("nvidia")},
+				{Name: "model", StringValue: ptr("dgxa100")},
+			},
+			Resources: sharedGroupToResources(model.NamedResources.SharedLimits[0]),
+			Devices:   devices,
+		},
+	}, nil
+}
+
+func instanceToDevice(instance newresourceapi.NamedResourcesInstance) api.Device {
+	device := api.Device{
+		Name:       instance.Name,
+		Attributes: attributesToAttributes(instance.Attributes),
+	}
+
+	if len(instance.Resources) > 0 {
+		device.Requests = sharedGroupToRequests(instance.Resources[0])
+		//device.Resources = sharedGroupToResources(instance.Resources[0])
+	}
+
+	return device
+}
+
+func attributesToAttributes(attrs []resourceapi.NamedResourcesAttribute) []api.Attribute {
+	var attributes []api.Attribute
+
+	for _, a := range attrs {
+		if a.QuantityValue != nil {
+			attributes = append(attributes, api.Attribute{
+				Name:          a.Name,
+				QuantityValue: a.QuantityValue,
+			})
+		} else if a.StringValue != nil {
+			attributes = append(attributes, api.Attribute{
+				Name:        a.Name,
+				StringValue: a.StringValue,
+			})
+		} else if a.IntValue != nil {
+			attributes = append(attributes, api.Attribute{
+				Name:     a.Name,
+				IntValue: ptr(int(*a.IntValue)),
+			})
+		} else if a.VersionValue != nil {
+			attributes = append(attributes, api.Attribute{
+				Name:        a.Name,
+				SemVerValue: ptr(api.SemVer(*a.VersionValue)),
+			})
+		}
+		// don't convert ones not supported in the prototype
+	}
+
+	return attributes
+}
+
+func sharedGroupToResources(group newresourceapi.NamedResourcesSharedResourceGroup) []api.ResourceCapacity {
+	var resources []api.ResourceCapacity
+
+	for _, item := range group.Items {
+		if item.QuantityValue != nil {
+			resources = append(resources, api.ResourceCapacity{
+				Name:     item.Name,
+				Capacity: *item.QuantityValue,
+			})
+		} else if item.IntRangeValue != nil {
+			// sorry, unrolling these intranges to avoid additional types
+			// beyond Quantity. Assumes max range 0-15
```

Comment:
I think this is fine for now. I think the ranges can help the scheduler to make better decisions to avoid fragmentation, but they are not necessarily *required* for correctness. If / when we decide to extend the model to include these, the scheduler can take advantage of them then.

Reply:
Unless we add a level of indirection in the `ResourceCapacity` type (or wrap that in something else) we won't be able to add such an extension later.